File: sunghyun/PGS/더 맵게.java
class Solution {
    class MinHeap { //최소힙 클래스 선언
        private int[] heap; // 힙 선언
        private int size; // 크기 선언
        
        public MinHeap(int maxSize) {
            heap = new int[maxSize + 1]; // 
            size = 0;
        }
        
        private int parent(int pos) { return pos / 2; } // 부모 노드
        private int leftChild(int pos) { return 2 * pos; } // 왼쪽 자식 노드
        private int rightChild(int pos) { return (2 * pos) + 1; } // 오른쪽 자식 노드
        
        private void swap(int pos1, int pos2) {
            int temp = heap[pos1];
            heap[pos1] = heap[pos2];
            heap[pos2] = temp;
        } //swap 메서드 (새로운 값이 추가될 때 트리가 힙이 되도록 자식노드와 부모노드를 바꾸기 위함)
        
        public void insert(int element) {
            heap[++size] = element;
            int current = size;
            
            while (current > 1 && heap[current] < heap[parent(current)]) {
                swap(current, parent(current));
                current = parent(current);
            }
        }
        
        public int remove() {
            if (size == 0) return -1;
            
            int popped = heap[1];
            heap[1] = heap[size--];
            
            int pos = 1;
            while (true) {
                int smallest = pos;
                int left = leftChild(pos);
                int right = rightChild(pos);
                
                if (left <= size && heap[left] < heap[smallest])
                    smallest = left;
                if (right <= size && heap[right] < heap[smallest])
                    smallest = right;
                
                if (smallest == pos) break;
                
                swap(pos, smallest);
                pos = smallest;
            }
            
            return popped;
        }
        
        public int peek() {
            if (size == 0) return -1;
            return heap[1];
        }
        
        public int getSize() {
            return size;
        }
    }
    
    public int solution(int[] scoville, int K) {
        int answer = 0;
        
        // 최소 힙 생성
        MinHeap minHeap = new MinHeap(scoville.length);
        
        // 모든 스코빌 지수를 힙에 삽입
        for (int s : scoville) {
            minHeap.insert(s);
        }
        
        // 가장 맵지 않은 음식의 스코빌 지수가 K보다 작고, 힙에 원소가 2개 이상 있을 때 계속 반복
        while (minHeap.peek() < K && minHeap.getSize() >= 2) {
            int first = minHeap.remove();
            int second = minHeap.remove();
            
            int newScoville = first + (second * 2);
            minHeap.insert(newScoville);
            
            answer++;
        }
        
        // 모든 음식의 스코빌 지수를 K 이상으로 만들 수 없는 경우
        if (minHeap.peek() < K) {
            return -1;
        }
        
        return answer;
    }
}

File: jiwook/Heap/더 맵게.py
import heapq
def solution(scoville, K):

    count = 0
    
    # heapq.heapify(): 리스트를 최소 힙으로 변환
    # 최소 힙: 부모 노드가 자식 노드보다 작은 값을 가지는 이진 트리
    heapq.heapify(scoville)
    
    # while 루프 조건:
    # 1. len(scoville) > 1: 최소 2개의 음식이 있어야 섞을 수 있음
    # 2. scoville[0] < K: 가장 작은 스코빌 지수가 K보다 작을 때만 계속 섞음
    while len(scoville) > 1 and scoville[0] < K:
       
        count += 1
        
        # heappop(): 힙에서 가장 작은 값을 제거하고 반환
        # first: 가장 맵지 않은 음식의 스코빌 지수
        first = heapq.heappop(scoville)
        # second: 두 번째로 맵지 않은 음식의 스코빌 지수
        second = heapq.heappop(scoville)
        
        # 문제에서 주어진 공식대로 두 음식을 섞어서 새로운 스코빌 지수 계산
        # heappush(): 계산된 새로운 스코빌 지수를 힙에 추가
        heapq.heappush(scoville, first + (second * 2))
    
    # 모든 반복이 끝난 후:
    # - 힙의 root(가장 작은 값)가 K보다 작으면 목표를 달성할 수 없으므로 -1 반환
    # - 그렇지 않으면 섞은 횟수(count) 반환
    return -1 if scoville[0] < K else count